[WFCORE-6135] TestSuite - NPE fastfail and safe tearDown

boris-unckel opened the pull request:
Fixes https://issues.redhat.com/browse/WFCORE-6135

yersan reviewed (changes requested):
Hello @boris-unckel , most of the changes here are changing how the cleanup tasks are executed. It is expected that those cleanup tasks are executed independently of whether they cannot be executed because of null variables. 

In such a hypothetical case, we should fail, however by applying these changes we no longer fail, we will silently continue without checking whether the cleanup routine is executed, making the test weaker.

Could you also describe a bit more how `AdminOnlyModeTestCase` would help you with the https://issues.redhat.com/browse/WFCORE-6135 description?

boris-unckel commented:
@yersan Thanks for the detailed review. I'm going to fix the issues tomorrow. I want to bring one thing to your attention: BeforeClass and AfterClass are independent. Regardless whether BeforeClass breaks or not, AfterClass is called. You can't assume from the broken state of BeforeClass to the beginning of AfterClass. 

I agree that things which cause an error today should not be silent: This change will cause failures (as term of JUnit) then.

boris-unckel commented:
> In such a hypothetical case, we should fail, however by applying these changes we no longer fail, we will silently continue without checking whether the cleanup routine is executed, making the test weaker.

Yes, agreed: A test failing today with a late error shall fail with an early failure after the change. The test keeps strong, respecting Before/After behaviour and making implicit assumptions explicit.

> 
> Could you also describe a bit more how `AdminOnlyModeTestCase` would help you with the https://issues.redhat.com/browse/WFCORE-6135 description?

I have updated the description, adding the current situation, describing the JUnit behaviour. I have changed the solution proposal, to respect the discussed issues.

boris-unckel commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/OperationTimeoutTestCase.java:
```diff
@@ -195,14 +195,20 @@ public static void tearDownDomain() throws Exception {
                 PathElement.pathElement(PROFILE, "default"),
                 PathElement.pathElement(SUBSYSTEM, BlockerExtension.SUBSYSTEM_NAME)));
-        executeForResult(safeTimeout(removeSubsystem), primaryClient);
+        if (primaryClient != null) {
```
BeforeClass and AfterClass are independent, if primaryClient creation fails in BeforeClass, the AfterClass will still be called. For the next comments I'm going to refer this with "Before/After independence".
NullCheck turned to assertion.

boris-unckel commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/events/JmxControlledStateNotificationsTestCase.java:
```diff
@@ -84,10 +84,18 @@ public static void setupClass() throws Exception {
-        testSupport.close();
-        testSupport = null;
-        domainPrimaryLifecycleUtil = null;
+        if (domainPrimaryLifecycleUtil != null) {
```
"Before/After independence". NullCheck turned to assertion.

boris-unckel commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/management/HttpManagementConstantHeadersTestCase.java:
```diff
@@ -279,7 +285,7 @@ private static ModelNode createHeaderMapping(final String path, final Map<String
    private void reload() throws IOException, TimeoutException, InterruptedException {
       domainPrimaryLifecycleUtil.reload("primary", null, true);
-
+      assertNotNull(domainSecondaryLifecycleUtil);
```
In general I agree and have done so, but in this case domainSecondaryLifecycleUtil is not in use for every test.

boris-unckel commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/management/cli/ManagementOpTimeoutTestCase.java:
```diff
@@ -68,18 +69,25 @@ public class ManagementOpTimeoutTestCase extends AbstractCliTestBase {
     public void closeServer() throws Exception {
-        cli.sendLine("/subsystem=blocker-test:remove", true);
-        cli.sendLine("/extension=org.wildfly.extension.blocker-test:remove");
+        if (cli != null) {
```
Fixed with assertion.

boris-unckel replied:
Thanks for the explanation. I have changed it to assertions.

boris-unckel commented on testsuite/rbac/src/test/java/org/jboss/as/test/integration/mgmt/access/AbstractPropertiesRoleMappingTestCase.java:
```diff
@@ -63,15 +64,30 @@ abstract class AbstractPropertiesRoleMappingTestCase extends AbstractRbacTestCas
-        managementClient.close();
-        container.stop();
+        try {
+            if (managementClient != null) {
```
Fixed with assertion.

boris-unckel commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/AbstractConfigurationChangesTestCase.java:
```diff
@@ -85,14 +87,20 @@ public static void setupDomain() throws Exception {
+            domainPrimaryLifecycleUtil = null;
+            domainSecondaryLifecycleUtil = null;
+            testSupport = null;
+        }
```
"Before/After independence". NullCheck turned to assertion.

boris-unckel commented on testsuite/domain/src/test/java/org/jboss/as/test/integration/domain/DomainControllerMigrationTestCase.java:
```diff
@@ -109,8 +109,9 @@ public static void shutdownDomain() throws IOException {
-        Assert.assertTrue(jarFile.delete());
+        if (jarFile != null) {
+            Assert.assertTrue(jarFile.delete());
+        }
```
Fixed with assertion.

boris-unckel commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/auditlog/AuditLogBootingLogTestCase.java:
```diff
@@ -145,6 +145,7 @@ public void beforeTest() throws Exception {
 
         // Start the server
         container.start();
+        Assert.assertTrue(container.isStarted()); // if container is not started, we get a NPE in container.getClient()
```
Fixed - I have removed it from the ```@Before``` cases.

boris-unckel commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/logging/LoggingPreferencesTestCase.java:
```diff
@@ -75,7 +75,9 @@ public class LoggingPreferencesTestCase extends AbstractLoggingTestCase {
     @Before
     public void prepareContainer() throws Exception {
         // Start the container
+        Assert.assertNotNull(container);
         container.start();
```
This is again the preference for failure instead of error for tests, making state assumptions explicit.

boris-unckel commented on testsuite/manualmode/src/test/java/org/jboss/as/test/manualmode/management/cli/ManagementOpTimeoutTestCase.java:
```diff
@@ -68,18 +69,25 @@ public class ManagementOpTimeoutTestCase extends AbstractCliTestBase {
         container.start();
+        Assert.assertTrue(container.isStarted()); // if container is not started, we get a NPE in container.getClient()
         initCLI();
+        Assert.assertNotNull(cli); // prevent subsequent NPEs in tests
```
Fixed, removed.

yersan commented:
@boris-unckel , thanks!
